lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskCloseHandler.cs

```diff
@@ -0,0 +1,108 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Text;
+using System.Threading;
+using Org.Apache.REEF.Common.Tasks.Events;
+using Org.Apache.REEF.IMRU.OnREEF.Driver;
+using Org.Apache.REEF.IMRU.OnREEF.Parameters;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.IMRUTasks
+{
+    internal sealed class TaskCloseHandler
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(TaskCloseHandler));
+
+        /// <summary>
+        /// When receiving a close event, this variable is set to 1. At the beginning of each task iteration,
+        /// if this variable is set to 1, the task will break from the loop and return from the Call() method.
+        /// </summary>
+        private long _shouldCloseTask = 0;
+
+        /// <summary>
+        /// Before the task is returned, this variable is set to 1.
+        /// Close handler will check this variable to decide if it needs to throw an exception.
+        /// </summary>
+        private long _isTaskStopped = 0;
+
+        /// <summary>
+        /// Waiting time for the task to close by itself
+        /// </summary>
+        private readonly int _enforceCloseTimeoutMilliseconds;
+
+        /// <summary>
+        /// An event that will wait in close handler until it is either signaled from Call method or timeout.
+        /// </summary>
+        private readonly ManualResetEventSlim _waitToCloseEvent = new ManualResetEventSlim(false);
+
+        /// <summary>
+        /// Handle task close event and manage the states, wait/signal when closing the task
+        /// </summary>
+        /// <param name="enforceCloseTimeoutMilliseconds">Timeout in milliseconds to enforce the task to close if receiving task close event</param>
+        [Inject]
+        private TaskCloseHandler([Parameter(typeof(EnforceCloseTimeoutMilliseconds))] int enforceCloseTimeoutMilliseconds)
+        {
+            _enforceCloseTimeoutMilliseconds = enforceCloseTimeoutMilliseconds;
+        }
+
+        /// <summary>
+        /// Handle Task close event.
+        /// Set _shouldCloseTask to 1 so that to inform the task to stop at the end of the current iteration.
+        /// Then waiting for the signal from Call method. Either it is signaled or after _enforceCloseTimeoutMilliseconds,
+        /// If the closed event is sent from driver, checks if the task has been stopped. If not, throw IMRUTaskSystemException 
+        /// to enforce the task to stop.
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        internal void HandleEvent(ICloseEvent closeEvent)
+        {
+            Interlocked.Exchange(ref _shouldCloseTask, 1);
+            _waitToCloseEvent.Wait(TimeSpan.FromMilliseconds(_enforceCloseTimeoutMilliseconds));
+
+            if (closeEvent.Value.IsPresent())
```

Comment:
It is possible the close handler is called through Context.Dispose(). In that case, the value is not present. I would expect it returns without throw any exception in that case.